presto-docs/src/main/sphinx/functions/window.rst

```diff
@@ -92,32 +92,36 @@ Value Functions
 
 .. function:: first_value(x) -> [same as input]
 
-    Returns the first value of the window.
+    Returns the first value of the window frame.
 
 .. function:: last_value(x) -> [same as input]
 
-    Returns the last value of the window.
+    Returns the last value of the window frame.
 
 .. function:: nth_value(x, offset) -> [same as input]
 
-    Returns the value at the specified offset from beginning the window.
+    Returns the value at the specified offset from the beginning of the window frame.
     Offsets start at ``1``. The offset can be any scalar
     expression.  If the offset is null or greater than the number of values in
-    the window, null is returned.  It is an error for the offset to be zero or
+    the window frame, ``null`` is returned.  It is an error for the offset to be zero or
     negative.
 
 .. function:: lead(x[, offset [, default_value]]) -> [same as input]
 
-    Returns the value at ``offset`` rows after the current row in the window.
+    Returns the value at ``offset`` rows after the current row in the window partition.
     Offsets start at ``0``, which is the current row. The
     offset can be any scalar expression.  The default ``offset`` is ``1``. If the
-    offset is null or larger than the window, the ``default_value`` is returned,
-    or if it is not specified ``null`` is returned.
+    offset is null, ``null`` is returned. If the offset refers to a row that is not
+    within the partition, the ``default_value`` is returned, or if it is not specified
+    ``null`` is returned.
+    If the window frame is specified, it is ignored by the :func:`lead` function.
 
 .. function:: lag(x[, offset [, default_value]]) -> [same as input]
 
-    Returns the value at ``offset`` rows before the current row in the window
-    Offsets start at ``0``, which is the current row.  The
+    Returns the value at ``offset`` rows before the current row in the window partition.
+    Offsets start at ``0``, which is the current row. The
     offset can be any scalar expression.  The default ``offset`` is ``1``. If the
-    offset is null or larger than the window, the ``default_value`` is returned,
-    or if it is not specified ``null`` is returned.
+    offset is null, ``null`` is returned. If the offset refers to a row that is not
+    within the partition, the ``default_value`` is returned, or if it is not specified
+    ``null`` is returned.
+    If the window frame is specified, it is ignored by the :func:`lag` function.
```

Comment:
That doesn't seem consistent with what the spec says:

```
b) If <window function type> is <lead or lag function>, then:
    i) Let OFFSET be the value of OFF and let DEFAULT be the value of VE2.
    ii) Let T be the collection of rows in the window frame of the current row 
        defined by WDX, as specified by the General Rules of Subclause 7.15, “<window clause>”.
```

Reply:
That is actually how our implementation works. Postgres does the same. Postgres documentation clearly says that while `first_value`, `last_value` and `nth_value` apply to window frame, `lag` and `lead` are computed with reference to the whole partition.
Could you please provide more spec for `lag`/`lead`?

Reply:
```
b) If <window function type> is <lead or lag function>, then:
  i) Let OFFSET be the value of OFF and let DEFAULT be the value of VE2.
  ii) Let T be the collection of rows in the window frame of the current row defined 
      by WDX, as specified by the General Rules of Subclause 7.15, “<window clause>”.
  iii) If LEAD is specified, then: 
    1) Case:
      A) If NTREAT is RESPECT NULLS, then let TX be the sequence of values that is 
         the result of applying VE1 to each row of T that follows the current row, ordered 
         according to the window ordering of WDX.
      B) Otherwise, let TX be the sequence of values that is the result of applying VE1 to 
         each row of T that follows the current row and eliminating null values, ordered 
         according to the window ordering of WDX.
    2) Let n be the number of values in TX.
    3) Case:
      A) If OFFSET > n, then the value of <window function> is DEFAULT.
      B) If OFFSET = 0 (zero), then the value of <window function> is the value
         of VE1 evaluated for the current row.
      C) Otherwise, the value of <window function> is the m-th value of TX, 
         where m = OFFSET.
```

The important part is where it says:
> Let `T` be the collection of rows in the window frame of the current row defined 
      by `WDX`, as specified by the General Rules of Subclause 7.15, “`<window clause>`”.

Reply:
Ah, but here's another important bit:

```
a) If <ntile function>, <lead or lag function>, RANK or DENSE_RANK is specified, then the window ordering clause WOC of WDX shall be present.
b) The window framing clause of WDX shall not be present.
```

So, for those functions, the `ORDER BY` clause is *required* and window framing clause is *illegal* (and should cause the query to fail).

Reply:
Right, so the current implementation is not consistent with the spec. However, it is consistent with PG :/ What should be done about it? We may want to decide before `RESPECT / IGNORE NULLS` is added.